### DISCUSSION TITLE
ECS?

### DISCUSSION THREAD

**Discussion by test3211234** [vizia/vizia#512] [Questions]
What are each of the Entity, Component and Systems in this library?

**Comment by geom3trik** [comment]
From ARCHITECTURE.md:

Each view in Vizia is assigned a generational `Entity` ID when created.

The properties of a view are not stored in the tree itself but instead in separate stores in the `Style` struct. The entity ID is used to set/get style properties (i.e. components) which are stored in custom sparse set storage types.

A series of systems are used to update the state of the application on each update cycle:
- Event Manager - Routes events in the event queue to Models and Views and calls the view/model `event()` method.
- Binding System - Queries model data for changes and rebuilds binding views.
- Style System - Links entities to shared style data (from CSS) and applies any style property inheritance.
- Image System - Loads any unloaded images and removes any unused images.
- Animation System - Applies any animations to style properties.
- Layout System - Determines the size and position of views.
- Accessibility System - Constructs or updates the accessibility tree and calls the view `accessibility()` method.
- Draw System - Draws the views to the main window and calls the view `draw()` method.

**Reply by test3211234** [reply]
Right, I was mainly confused about Components? Searching "component" in the code leads to nothing. How are components made?

**Reply by rhelmot** [reply]
vizia doesn't have a "true ecs" per se, there are no components that can be arbitrarily attached to entities. Instead, vizia defines a huge number of pre-set "components" in the form of indexing large trees containing known data structures with predefined meanings using the entity identifiers. For example, the style system maintains a mapping of entity to style directives, entity to style cache, etc.

**Reply by test3211234** [reply]
The vizia_style folder is massive. Is there any way I can simply see how you implemented the "ECS" system? Maybe the main files?

**Reply by rhelmot** [reply]
Like I said, there is no real ECS, just an informal system of tracking data related to entities. Is there a more specific question you'd like answered?

**Reply by test3211234** [reply]
Can you explain the TourStep and TourDirection stuff? I would think iterating the tree is as simple as going through the children, and the children's children etc.

**Reply by 0HyperCube** [reply]
@test3211234 iterating through a tree can be done by recursively iterating through each node's children. However in order to create a rust iterator, it is necessary to use iterative code (since the call stack is not preserved since rust doesn't have [generators](https://github.com/samsartor/lang-team/blob/master/src/design_notes/general_coroutines.md)).

TourStep and TourDirection are used to implement that iterative tree traversal.
